### ISSUE TITLE
@reloadscript doesn't apply mapflags in instances

### ISSUE THREAD

**Issue by secretdataz** [rathena/rathena#798]
While using `@reloadscript` reset instances properly, mapflags won't apply.

Before reloadscript : 
![image](https://cloud.githubusercontent.com/assets/2021243/11528973/93b69448-991c-11e5-8187-133ee2a846cb.png)

After reloadscript : 
![image](https://cloud.githubusercontent.com/assets/2021243/11528984/a9768fa4-991c-11e5-92ee-1eb6f1cefe8c.png)

Sorry for Thai language in screenshots.

Using commit : c7d7371

**Comment by secretdataz** [comment]
Client date : 2015-05-13
However, I don't think this has anything to do with the client.

**Comment by julia40124009** [comment]
confirm
same issue

**Comment by cydh** [comment]
@secretdataz @julia40124009, can u guys test this gist https://gist.github.com/cydh/87d4ce977e3830a36644 ?

**Comment by secretdataz** [comment]
@cydh Your fix works like a charm. Applying it on 70eb71d fixes this issue.  
Thanks ;)  
![image](https://cloud.githubusercontent.com/assets/2021243/11837128/947af5f4-a411-11e5-9dce-ef6932037212.png)

**Comment by Lemongrass3110** [comment]
@cydh your fix is good.
The problem might be that it could interfere with the instance's script logic. For example if the instance sets any mapflag that is not set by default or removes any mapflag that is set by default. Therefore the right way would be as you said, setting them on OnInstanceInit. Even that could cause unknown problems. Another problem that might occur is for cell options(like setting a cell walkable or not).
I think the best way to ensure all instances are clean would be to destroy all active instances on reloadscript and warp all players in instances to their savepoints, because reloadscript should mimic a server restart and instance states are not saved if the server is restarted.

**Comment by secretdataz** [comment]
While I agree with @Lemongrass3110 , I think we should make it optional for the server owners to decide if they want the "Safe Approach" or the old way. Maybe add it to map server config?.
Because the instance reset on `@reloadscript` is so handy when developing a custom instance or editing one. Additionally, some script devs won't need that much stability when they are testing scripts in their local server.

**Comment by cydh** [comment]
@Lemongrass3110 seems ur logic's flow is broken somewhere.

user: I have map M with a b c  .. x y z mapflags, I made instance, the map is M. Oh the M's mapflags are copied on instance, nice. ... later...Uhhh, I screw the script, need to `@reloadscript`. What? Player can memo this map now??!!

If u (alone or rA devs) don't want map's 'original' mapflags are being copied, never do it at first place. /ho

**Comment by cydh** [comment]
@secretdataz better u forget my diff, just use trick I told. This won't be fixed "easily" in current rAthena mapflag system.

**Comment by secretdataz** [comment]
Well, I don't really have any problem with this. I just report a bug I found while implementing Infinite Space instance. Anything else is up to the rA dev team ;)
And, yeah, I already migrated to `OnInstanceInit`. The script dev team may also have to migrate all official instance scripts to this system to avoid this bug.

**Comment by Lemongrass3110** [comment]
@cydh how do we want to deal with this one now? I understand your point, I just pointed out, that there might be some mapflags that were set by the script for example "noteleport" if someone does not want they players to warp away during some steps of the instance script but enabling it back later on.
It would be nice if you could join us on discord so we can talk about this together with @aleos89 and the others.

**Comment by Grantea** [comment]
I don't if this is still an issue but the same thing happened to me. 

Here's my fix:
```
     OnInit:
		addrid(0);
		.@instance_map$ = "";
		.@instance_map$ = instance_mapname("<map name>", instance_id());
		if(.@instance_map$ == "") end;
        // I'm checking just one because I'm lazy, but you can check all mapflags you're using if you want.
		if(!getmapflag(.@instance_map$, mf_nowarp))  
		{
			setmapflag .@instance_map$, mf_monster_noteleport;
			setmapflag .@instance_map$, mf_nowarp;
			setmapflag .@instance_map$, mf_nowarpto;
			setmapflag .@instance_map$, mf_nomemo;
			setmapflag .@instance_map$, mf_noteleport;
			setmapflag .@instance_map$, mf_nosave;
			setmapflag .@instance_map$, mf_nopenalty;
		}
		end;	
```

**Comment by darknessfmy** [comment]
@cydh hi, you diff no work in actual rA..

**Comment by secretdataz** [comment]
Fixed in e66dc4f.
